### PR TITLE
Make tests parallel.

### PR DESCRIPTION
**Pull request by egonelbre**
On my local machine the tests improved from 99s to 50s.

I currently used a batch approach to add t.Parallel. Not all packages had significant benefit, but tried to replace as many to have `t.Parallel` used consistently.

There are places where I didn't replace, because they require some more effort to replace.

All commits are individually reviewable - or if the non-significant improvements aren't necessary for packages, then we can remove those commits.

### CODE REVIEW

**Comment by egonelbre** [comment]
Note, I just realized that I edited some generated code, will fix in a bit.

**Comment by egonelbre** [comment]
Comparison of test runtime: https://gist.github.com/egonelbre/720ee94004ef37d9926627c6e4fed02d

File: graph/community/bisect_test.go [hunk 1]
```diff
@@ -137,6 +137,7 @@ func ExampleProfile_multiplex() {
 }
 
 func TestProfileUndirected(t *testing.T) {
+	t.Parallel()
```
Comment (kortschak) [inline graph/community/bisect_test.go]: These tests are not safe here. You can file an issue for me to look into this later, but for the moment they should probably be backed out.

Comment (egonelbre) [reply graph/community/bisect_test.go]: Yup, I'll remove that commit.

I'm also wondering whether to keep or remove changes in packages that don't have significant impact?

File: diff/fd/hessian_test.go [hunk 2]
```diff
@@ -16,12 +16,14 @@ type HessianTester interface {
...
-var hessianTestCases = []struct {
+type hessianTestCase struct {
 	h        HessianTester
 	x        []float64
 	settings *Settings
 	tol      float64
-}{
+}
+
+var _hessianTestCases = []hessianTestCase{
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply diff/fd/hessian_test.go]: Currently the tests are modifying the test cases, so this change was the smallest fix for this without duplicating code.

File: unit/constant/generate_constants.go [hunk 3]
```diff
@@ -279,6 +279,7 @@ import (
 )
 
 func Test{{.Name}}Format(t *testing.T) {
+	t.Parallel();
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply unit/constant/generate_constants.go]: ah, missed the generated code changes

File: unit/generate_unit.go [hunk 4]
```diff
@@ -615,6 +616,7 @@ func Test{{.DimensionName}}(t *testing.T) {
 }
 
 func Test{{.DimensionName}}Format(t *testing.T) {
+	t.Parallel();
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply unit/generate_unit.go]: done

File: diff/fd/jacobian_test.go [hunk 5]
```diff
@@ -56,7 +56,8 @@ func vecFunc43Jac(jac *mat.Dense, x []float64) {
 }
 
 func TestJacobian(t *testing.T) {
-	rand.Seed(1)
+	t.Parallel()
+	rng := rand.New(rand.NewSource(1))
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply diff/fd/jacobian_test.go]: done

File: diff/fd/hessian_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply diff/fd/hessian_test.go]: Ah yes. Thanks

**Comment by kortschak** [comment]
Looks like there's a new failure that is sporadic in mat/TestDenseMul. I'll look in to it.

**Comment by egonelbre** [comment]
@kortschak after reviewing a bit, this suggests to me that some other test either modifies the testMatrices. Such flakiness could also happen with rand giving values that the code doesn't know how to handle, however, in this case it doesn't look like rand is used in TestDenseMul.

**Comment by kortschak** [comment]
Yes, those were my suspicions too. The `makeRandOf` case should be fixed.

**Comment by egonelbre** [comment]
With https://gist.github.com/egonelbre/aeec9f26c7d8e62ce64eebc78d543c29 and `go test -count 100 -failfast .` it fails quite fast for me.

**Comment by kortschak** [comment]
I've sent #1291 to address this.

**Comment by kortschak** [comment]
BTW It would be good to have a look at one of those failing cases, just to make sure we're not papering over something.

**Comment by egonelbre** [comment]
For example trying to debug the it further it's getting:
```
    list_test.go:1313: Answer mismatch non-zero receiver: *mat.Dense.Mul(*mat.Dense, *mat.SymDense), sizes: struct { ar int; ac int; br int; bc int }{ar:2, ac:2, br:2, bc:2}, atrans true, btrans false
        A=
            ⎡ 0.5591351826376594  0.32575892467275946⎤
            ⎣0.29137243197115714   0.2868964560092573⎦ 
        B=
            ⎡0.018567058293321392    0.3077178510998605⎤
            ⎣  0.3077178510998605   0.20718225824279746⎦
        neverZero before=
            ⎡0.9172763622957576  0.5563345111752376⎤
            ⎣0.9573264232865906  0.7287975046717311⎦
        Want=
            ⎡0.11062333180678319  0.23954734653203624⎤
            ⎣0.09369308986080961   0.1491003542737581⎦
        neverZero=
            ⎡0.11062333180678319  0.23954734653203624⎤
            ⎣                NaN   0.1491003542737581⎦
FAIL
```

**Comment by kortschak** [comment]
It's not clear to me what is happening here, replicating that in a standalone does the correct thing.

https://play.golang.org/p/vAvxlLs33Kd

**Comment by egonelbre** [comment]
@kortschak yeah, it's super weird. Could some test be corrupting stacks? Alternatively, is this a compiler bug? I cannot explain where the NaN comes from.

**Comment by egonelbre** [comment]
Btw, are you able to reproduce locally?

**Comment by kortschak** [comment]
I will take a look tomorrow.

**Comment by kortschak** [comment]
OK, I had a quick look and I can reproduce it. I can also stop it by inactivating the workspace pool for `*Dense`; so something is contaminating that. I will look into it further tomorrow.

**Comment by kortschak** [comment]
OK. It looks like its contamination of the pool, probably with the NaN and that the gonum implementation of Dsymm is incorrectly handling the case when the beta parameter is zero. This can be shown by changing the `false` to `true` in the two calls to `getWorkspace` in the `*SymDense` case of type switch in the `Mul` method with the changes you have above. Obviously this is not the correct fix  since it does double work.

I'll have a look at `Dsymm` later today.

**Pull request merged**